scripts/add_electricity.py

```diff
@@ -0,0 +1,262 @@
+        return pd.Series(
+            dict(capital_cost=capital_cost, marginal_cost=0.0, co2_emissions=0.0)
+        )
+    
+    max_hours = elec_config["max_hours"]           
+    costs.loc["battery"] = costs_for_storage(                      
+            costs.loc["lithium"],                            #line 119 in file costs.csv' which was battery storage was modified into lithium (same values left)
+            costs.loc["battery inverter"],
+            max_hours=max_hours["battery"],
+    )
+    max_hours = elec_config["max_hours"]
+    costs.loc["battery"] = costs_for_storage(
+            costs.loc["lead acid"],                          #line 120 in file 'costs.csv' which was battery storage was modified into lithium (same values left)
+            costs.loc["battery inverter"],
+            max_hours=max_hours["battery"],
+    )
+    
+    costs.loc["H2"] = costs_for_storage(
+        costs.loc["hydrogen storage"],
+        costs.loc["fuel cell"],
+        costs.loc["electrolysis"],
+        max_hours=max_hours["H2"],
+    )
+
+    for attr in ("marginal_cost", "capital_cost"):
+        overwrites = config.get(attr)
+        if overwrites is not None:
+            overwrites = pd.Series(overwrites)
+            costs.loc[overwrites.index, attr] = overwrites
+
+    return costs
+
+def attach_wind_and_solar(n, costs, input_profiles, tech_modelling, extendable_carriers):
+
+    _add_missing_carriers_from_costs(n, costs, tech_modelling)
+
+    for tech in tech_modelling:
+       
+        # with xr.open_dataset(f"renewable_profiles/profile_{tech}.nc") as ds:
+        with xr.open_dataset(getattr(snakemake.input, "profile_" + tech)) as ds:
+            
+            if ds.indexes["bus"].empty:
+                continue   
+
+            suptech = tech.split("-", 2)[0]
+          
+            n.madd(
+            "Generator",
+            ds.indexes["bus"],
+            " " + tech,
+            bus=["onebus"],
+            carrier=tech,
+            p_nom_extendable=tech in extendable_carriers["Generator"],
+            p_nom_max=ds["p_nom_max"].to_pandas(), #look at the config 
+            weight=ds["weight"].to_pandas(),
+            marginal_cost=costs.at[suptech, "marginal_cost"],
+            capital_cost=costs.at[tech, "capital_cost"],
+            efficiency=costs.at[suptech, "efficiency"],
+            p_set=ds["profile"].transpose("time", "bus").to_pandas().reindex(n.snapshots),
+            p_max_pu=ds["profile"].transpose("time", "bus").to_pandas().reindex(n.snapshots),
+            )
+
+
+def attach_conventional_generators(n, tech_modelling, conventional_carriers):
+
+    for tech in tech_modelling:
+
+        buses_i = n.buses.index
+        n.madd(
+            "Generator",
+            buses_i ,
+            " " + tech, 
+            bus=["onebus"], 
+            carrier=tech,
+            p_nom_extendable=True,
+            p_nom_max=100,
+            marginal_cost=10, #random number (to be corrected)
```

Comment:
These parameters shall be taken using the costs dataframe that shall be given as input, not as global costants
Moreover, the bus=[...] shall be revise into bus=buses_i .

Reply:
The comment on bus= applies also in the other functions, as well as the global constants

Reply:
The bus correction was applied everywhere needed in the add electricity script. Furthermore, I changed the ""attach_conventional_generators" function, making it more similar to pypsa-earth. Now a diesel generator (from scratch)  is added starting from a csv file (powerplants.csv) and there is also the possibility to add existing powerplants like it is done in pypsa-earth just adding rows to that csv file (in future also the "build_powerplant" script coud be implemented and the function would not change consistently).